jpa/subsystem/src/main/java/org/jboss/as/jpa/transaction/TransactionUtil.java

```diff
@@ -137,47 +181,50 @@ public SessionSynchronization(EntityManager session, String scopedPuName) {
-                    if (ROOT_LOGGER.isDebugEnabled())
-                        ROOT_LOGGER.debugf(ignored, "ignoring error that occurred while closing EntityManager for %s (", scopedPuName);
-                }
-            }
-            // The TX reference to the entity manager, should be cleared by the TM
-
+            afterCompletionCalled = true;
+            safeCloseEntityManager();
         }
 
         /**
-         * AS7-6586 requires that the container avoid closing the EntityManager while the application
-         * may be using the EntityManager in a different thread.  If the transaction has been rolled
-         * back, will check if the current thread is the Arjuna transaction manager Reaper thread.  It is not
-         * safe to call EntityManager.close from the Reaper thread, so false is returned.
-         *
-         * TODO: switch to depend on JBTM-1556 instead of checking the current thread name.
-         *
-         * @param status of transaction.
-         * @return
+         * After the JTA transaction is ended (Synchronization.afterCompletion has been called) and
+         * the JTA transaction is no longer associated with application thread (application thread called
+         * transaction.rollback/commit/suspend), the entity manager can safely be closed.
          */
-        private boolean safeToClose(int status) {
-            if (Status.STATUS_COMMITTED != status) {
-                return !TxUtils.isTransactionManagerTimeoutThread();
+        private void safeCloseEntityManager() {
+            if ( afterCompletionCalled == true && transactionDisassociatedFromApplication == true) {
+                if (manager != null) {
+                    try {
+                        if (ROOT_LOGGER.isDebugEnabled())
+                            ROOT_LOGGER.debugf("%s: closing entity managersession", getEntityManagerDetails(manager, scopedPuName));
+                        manager.close();
+                    } catch (Exception ignored) {
+                        if (ROOT_LOGGER.isDebugEnabled())
+                            ROOT_LOGGER.debugf(ignored, "ignoring error that occurred while closing EntityManager for %s (", scopedPuName);
+                    }
+                    manager = null;
+                }
             }
+        }
 
-            return true;
+        @Override
+        public synchronized void onEvent(TransactionEvent transactionEvent) {
+            if (transactionEvent.getTypes().contains(EventType.ASSOCIATED)) {
+                // application thread is now associated with JTA transaction
+                transactionDisassociatedFromApplication = false;
+            }
+            else if(transactionEvent.getTypes().contains(EventType.DISASSOCIATING)) {
+                // application thread is no longer associated with JTA transaction
+                transactionDisassociatedFromApplication = true;
+            }
+            safeCloseEntityManager();
```

Comment:
Great questions, I should describe what is going on in this change.  I'll push a change that explains in a comment.  The Synchronization.afterCompletion() will eventually be called from either the application thread or a background (TM communication or TM timeout reaper) thread.  Eventually, both conditions are reached, that the application thread is disassociated from the TX *and* Synchronization.afterCompletion() has been called.  At which time its safe to close the EntityManager.  We do have lots of code that cause the transaction context to be suspended/resumed but unless the Synchronization.afterCompletion has also been called, the entity manager is not yet closed.  The previous solution only checked the TM thread name to see if it was the reaper thread and that doesn't work in all cases (e.g. when the background thread is a TM communications thread calling us because a remote JVM timed out the TX).
```
    /**
     * Using the TransactionListener, we can detect Synchronization.afterCompletion() being called from a
     * background thread, while the application may still be actively using the EntityManager.
     * We need to ensure that the background thread does not close the EntityManager while the application thread
     * is actively using it.
     *
     * We know when the application thread is associated with the transaction and can defer closing the EntityManager
     * until both conditions are met:
     *
     *   1. application is disassociated from transaction
     *   2. Synchronization.afterCompletion has been called
     *
     * See discussions for more details about how we arrived at using the TransactionListener:
     *     https://developer.jboss.org/message/919807
     *     https://developer.jboss.org/thread/252572
     */
'''